src/browser/modules/Editor/Monaco.tsx

```diff
@@ -41,193 +51,480 @@ import {
+      onChange = () => undefined,
+      onChangeLineCount = () => undefined,
+      onDisplayHelpKeys = () => undefined,
+      onExecute = () => undefined,
+      schema,
+      theme = LIGHT_THEME,
+      useDb
+    }: MonacoProps,
+    ref
+  ): JSX.Element => {
+    const editorRef = useRef<editor.IStandaloneCodeEditor | null>(null)
+    const monacoId = `monaco-${id}`
 
-    onContentUpdate()
+    useImperativeHandle(ref, () => ({
+      focus() {
+        editorRef.current?.focus()
+      },
+      getValue() {
+        return editorRef.current?.getValue() || ''
+      },
+      setValue(value: string) {
+        setValue(value)
+      },
+      resize(fillContainer = false, fixedHeight) {
+        resize(fillContainer, fixedHeight)
+      }
+    }))
 
-    editorRef.current?.onDidChangeModelContent(onContentUpdate)
+    const editorSupportRef = useRef<CypherEditorSupport | null>(null)
 
-    return () => {
-      editorRef.current?.dispose()
-    }
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [])
-
-  // Update theme when setting is changed
-  useEffect(() => {
-    editor.setTheme(theme)
-  }, [theme])
-
-  // Trigger update when multi statement setting is changed to update warnings
-  useEffect(() => {
-    onContentUpdate()
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [enableMultiStatementMode])
-
-  // Share current text with parent and add warnings
-  const updateCode = () => {
-    const text =
-      editorRef.current
-        ?.getModel()
-        ?.getLinesContent()
-        .join('\n') || ''
-
-    onChange(text)
-    addWarnings(parse(text).referencesListener.queriesAndCommands)
-  }
+    // The monaco render method does not redraw line numbers
+    // Getting and setting content and cursor is done to force a redraw
+    useEffect(() => {
+      const cursorPosition = editorRef?.current?.getPosition() as IPosition
+      editorRef.current?.setValue(editorRef.current?.getValue() || '')
+      editorRef.current?.setPosition(cursorPosition)
+    }, [useDb])
 
-  const debouncedUpdateCode = debounce(updateCode, 300)
+    // Create monaco instance, listen to text changes and destroy
+    useEffect(() => {
+      languages.register({ id: 'cypher' })
+      languages.setTokensProvider('cypher', new CypherTokensProvider())
 
-  // On each text change, clear warnings and reset countdown to adding warnings
-  const onContentUpdate = () => {
-    editor.setModelMarkers(
-      editorRef.current?.getModel() as editor.ITextModel,
-      monacoId,
-      []
-    )
-    debouncedUpdateCode()
-  }
+      editorSupportRef.current = new CypherEditorSupport(value)
+      if (schema) {
+        editorSupportRef.current.setSchema(schema)
+      }
 
-  const addWarnings = (
-    statements: { start: { line: number }; getText: () => string }[]
-  ) => {
-    if (!statements.length) return
+      languages.registerCompletionItemProvider('cypher', {
+        triggerCharacters: ['.', ':', '[', '(', '{'],
```

Comment:
Is there a way to avoid these triggering inside strings? Like in this example: match (n) where n.name = "oskar.

Reply:
It might be possible, I've added it to my todo for improving suggestions